numpy/__init__.pyi

```diff
@@ -311,52 +311,13 @@ from numpy.core.shape_base import (
     vstack as vstack,
 )
 
-# Add an object to `__all__` if their stubs are defined in an external file;
-# their stubs will not be recognized otherwise.
-# NOTE: This is redundant for objects defined within this file.
-__all__ = [
-    "linspace",
-    "logspace",
-    "geomspace",
-    "take",
-    "reshape",
-    "choose",
-    "repeat",
-    "put",
-    "swapaxes",
-    "transpose",
-    "partition",
-    "argpartition",
-    "sort",
-    "argsort",
-    "argmax",
-    "argmin",
-    "searchsorted",
-    "resize",
-    "squeeze",
-    "diagonal",
-    "trace",
-    "ravel",
-    "nonzero",
-    "shape",
-    "compress",
-    "clip",
-    "sum",
-    "all",
-    "any",
-    "cumsum",
-    "ptp",
-    "amax",
-    "amin",
-    "prod",
-    "cumprod",
-    "ndim",
-    "size",
-    "around",
-    "mean",
-    "std",
-    "var",
-]
+__all__: List[str]
+__path__: List[str]
+__version__: str
+__git_version__: str
```

Comment:
I'm not 100% sure if all these dunders are considered public or not (especially the bottom three). Are there any about annotating them?

Secondly, they could be marked with `Final` if they should be considered read-only.

Reply:
`__path__` is a python builtin that's added to all packages, like `__name__`.

Reply:
The annoying thing it is not defined for [`types.ModuleType`](https://github.com/python/typeshed/blob/a1a51f5aace0989301576c7f7edc7432bef6f5c9/stdlib/3/_importlib_modulespec.pyi#L35-L42) (as not all modules are packages), and by extension for `numpy`.
This is arguably an issue that should be solved in mypy though (or perhaps even typeshed).

Reply:
How about the likes of `np.__NUMPY_SETUP__`, `np.__deprecated_attrs__` and `np.__expired_functions__` though?
They're not (clearly) marked as private, but I'm not 100% sure whether or they're considered public.

Reply:
I would consider them private, maybe they should never have been a dunder as such.  I am least sure about `__NUMPY_SETUP__`, but frankly, I wonder if that will matter in a typing context.

Reply:
Yeah, for now I've just removed them (https://github.com/numpy/numpy/pull/18229/commits/7060d5e28f735199df6b465755ddd1d3bc7ed07e).
We can always consider revisiting this subject if the need arises.